### PR TITLE
refactor: drop joinWithDelim

### PR DESCRIPTION
**Pull request by aakropotkin**
- refactor: expropriate `concatStringsSep` from `nix`.
- refactor: replace existing `nix::concatStringsSep` with _our copy_.
- refactor: replace existing `joinWithDelim` calls with `concatStringsSep`.
- style: misc fixes to doc strings.
- test: do not rely on exact number of elements in some search tests.
- feat: add `EXPECT_FAIL` helper for tests.
- feat: add `std::vector<T>::operator==()` to `flox/core/util.hh`.
- fix: set paths to additional `pkgdb` executables in `devShell`.

This PR cherry picks parts of `aameen.scrape-rules` that are
_general improvements_ that can be merged independently.

### CODE REVIEW

File: pkgdb/include/flox/core/util.hh [hunk 1]
```diff
@@ -403,50 +427,48 @@ merge_vectors( const std::vector<T> & lower, const std::vector<T> & higher )
...
- * @a std::vector<T>.
+ * @brief Concatenate the given strings with a separator between
+ *        the elements.
  */
-template<typename T>
-[[nodiscard]] std::vector<std::optional<T>>
-vectorMapOptional( const std::vector<T> & orig )
+template<class Container>
+[[nodiscard]] std::string
+concatStringsSep( const std::string_view sep, const Container & strings )
```
Comment (ysndr) [inline pkgdb/include/flox/core/util.hh:445]: nonblocking: I still find the order of argument weird but

File: pkgdb/tests/search.bats
Comment (ysndr) [inline pkgdb/tests/search.bats:1]: nit: these change dont seem to be mentioned in the description
.. or i cant match them

**Review by ysndr — APPROVED** [review]
not sure if [pkgdb/tests/search.bats](https://github.com/flox/flox/pull/709/files#diff-833a08f04bb369f26931bda5074f5d147e07a3ceb1ae0a5a1824a60b11174012)
was supposed to be committed here or not, but otherwise LGTM

File: pkgdb/include/flox/core/util.hh (hunk 1 above)
Comment (aakropotkin) [reply pkgdb/include/flox/core/util.hh:445]: I just align with the existing `nix` one

File: pkgdb/tests/search.bats
Comment (aakropotkin) [reply pkgdb/tests/search.bats:1]: > test: do not rely on exact number of elements in some search tests.

Comment (ysndr) [reply pkgdb/tests/search.bats:1]: i see, thanks

File: pkgdb/include/flox/core/util.hh (hunk 1 above)
Comment (ysndr) [reply pkgdb/include/flox/core/util.hh:445]: ah the nix-lang one not the nix-cpp one then?
I guess makes sense to have _the list_ be the curryied argument there yes..

**Comment by aakropotkin** [comment]
This is breaking down parts of : https://github.com/flox/product/issues/607

**Pull request merged**